### PR TITLE
[IMP] eLearning: adds course and certification stat button on partner

### PR DESCRIPTION
**Pull request by pko-odoo**
Adds two stats button on res.partner form view. The first one will show all
the courses that the partner is following. The second one will show all
the certifications the partner has passed. When the partner is a company
all the courses followed by all the partners are shown. All the certifications
passed by all the partners are shown.

Task ID : 2034097

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website_slides_survey/models/res_partner.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    certification_ids = fields.One2many('slide.slide.partner', 'partner_id', domain=[('survey_quizz_passed', '=', True)])
+    certifications_count = fields.Integer('Succeeded Certifications Count', compute='_compute_certifications_count')
+    company_certification_ids = fields.One2many('slide.slide.partner', 'partner_id', compute='_compute_company_certifications_count')
+    company_certifications_count = fields.Integer('Succeeded Certification By Company Partners', compute='_compute_company_certifications_count')
+
+    def _compute_certifications_count(self):
+        self.certifications_count = len(self.certification_ids)
```
Comment (tde-banana-odoo) [inline addons/website_slides_survey/models/res_partner.py]: read_group_res = self.env['slide.slide.partner'].sudo().read_group(
  [('partner_id', 'in', self.ids), ('survey_quizz_passed', '=', True)],
  ['partner_id'],
  'partner_id')

See for example _compute_members_count in slide_channel.py

File: addons/website_slides_survey/models/res_partner.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+    def _compute_company_certifications_count(self):
+        self.company_certification_ids = self.env['slide.slide.partner'].search([
```
Comment (tde-banana-odoo) [inline addons/website_slides_survey/models/res_partner.py]: As this field is computed currently only for a form view we can just do sum(child.certification_count for child in record.child_ids)

File: addons/website_slides/models/res_partner.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    slide_channels_count = fields.Integer('Followed Course Count', compute='_compute_channels_count')
+    company_slide_channels_count = fields.Integer('Followed Course By Company Partners Count', compute='_compute_company_channels_count')
```
Comment (dbeguin) [inline addons/website_slides/models/res_partner.py]: I would name it slide_channels_company_count so that is is both 'prefixed' with slide_channels

File: addons/website_slides/views/res_partner_views.xml [hunk 4]
```diff
@@ -0,0 +1,72 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data>
+
+        <!--<record id="slide_channels_followed_by_company_partners_view" model="ir.ui.view">-->
```
Comment (dbeguin) [inline addons/website_slides/views/res_partner_views.xml]: to remove, same for all commented lines here.

File: addons/website_slides_survey/models/res_partner.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    certifications_count = fields.Integer('Succeeded Certifications Count', compute='_compute_certifications_count')
+    company_certifications_count = fields.Integer('Succeeded Certification By Company Partners', compute='_compute_company_certifications_count')
```
Comment (dbeguin) [inline addons/website_slides_survey/models/res_partner.py]: same here, name it : certifications_company_count

File: addons/website_slides_survey/views/res_partner_views.xml [hunk 6]
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data>
```
Comment (dbeguin) [inline addons/website_slides_survey/views/res_partner_views.xml]: to gain one indentation, you could directly do : 
`<odoo><data>`
on the same line

File: addons/website_slides/models/res_partner.py (hunk 3 above)
Comment (tde-banana-odoo) [reply addons/website_slides/models/res_partner.py]: +1

And maybe label to Company Course Count ?

File: addons/website_slides/models/res_partner.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    slide_channels_count = fields.Integer('Followed Course Count', compute='_compute_channels_count')
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/res_partner.py]: Simply Course Count ?

File: addons/website_slides/models/res_partner.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    company_slide_channels_count = fields.Integer('Followed Course By Company Partners Count', compute='_compute_company_channels_count')
+
+    def _compute_channels_count(self):
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/res_partner.py]: Once everyone agrees on computed field name, update the method name to _compute_field_name (_compute_slide_channel_count for example).

Don't forget the triggers, even for a non stored used in a form view it is better to give them

File: addons/website_slides/models/res_partner.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        read_group_res = self.env['slide.channel.partner'].sudo().read_group(
+            [('partner_id', 'in', self.ids)],
+            ['partner_id'], 'partner_id'
+        )
+        data = dict((res['partner_id'][0], res['partner_id_count']) for res in read_group_res)
+        for partner in self:
+            partner.slide_channels_count = data.get(partner.id, 0)
+
+    def _compute_company_channels_count(self):
+        search_res = self.env['slide.channel'].sudo().search(
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/res_partner.py]: You could simply do
```
for partner in self:
  if partner.is_company:
    partner.company_slide_channels_count = sum(child.slide_channels_count for child in partner.child_ids)
  else:
     partner.company_slide_channels_count = 0
```

Don't forget the triggers (is_company, child_ids.slide_channels_count)

File: addons/website_slides/models/res_partner.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+            [('partner_ids', 'in', self.child_ids.ids)]
+        )
+        self.company_slide_channels_count = len(search_res)
+
+    def action_redirect_to_followed_channels(self):
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/res_partner.py]: I would rename to action_view_courses a bit shorted (and action_view even if not in guidelines could become a guideline :p).

File: addons/website_slides/views/res_partner_views.xml [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+            <!--<field name="arch" type="xml">-->
+                <!--<tree string="Followed Courses By Partner" create="false" edit="false" delete="false">-->
+                    <!--<field name="name" string="Course"/>-->
+                    <!--<field name="channel_type"/>-->
+                    <!--<field name="completion"/>-->
+                <!--</tree>-->
+            <!--</field>-->
+        <!--</record>-->
+
+        <record id="res_partner_channels_action" model="ir.actions.act_window">
```
Comment (tde-banana-odoo) [inline addons/website_slides/views/res_partner_views.xml]: Just to follow guideline: res_partner_action_courses (or something similar)

File: addons/website_slides/views/res_partner_views.xml [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+            <field name="name">Followed Courses</field>
+            <field name="res_model">slide.channel</field>
+            <field name="view_mode">tree</field>
+            <!--<field name="view_id" ref="website_slides.slide_channels_followed_by_partners_view"/>-->
+        </record>
+
+        <record id="res_partner_view_form" model="ir.ui.view">
+            <field name="name">res.partner.course.stat.button</field>
```
Comment (tde-banana-odoo) [inline addons/website_slides/views/res_partner_views.xml]: Just to follow guideline: res.partner.view.form.inherit.slides

File: addons/website_slides/views/slide_channel_partner_views.xml [hunk 13]
```diff
@@ -32,5 +25,13 @@
                 </tree>
             </field>
         </record>
+
+        <record id="slide_channel_partner_action" model="ir.actions.act_window">
```
Comment (tde-banana-odoo) [inline addons/website_slides/views/slide_channel_partner_views.xml]: Why this change ? Normally defining manually the tree view_id is not necessary as we have only one default one.

File: addons/website_slides_survey/models/res_partner.py (hunk 5 above)
Comment (tde-banana-odoo) [reply addons/website_slides_survey/models/res_partner.py]: +1 see hereabove  same comments apply :)

File: addons/website_slides_survey/views/res_partner_views.xml (hunk 6 above)
Comment (tde-banana-odoo) [reply addons/website_slides_survey/views/res_partner_views.xml]: +1

File: addons/website_slides_survey/views/res_partner_views.xml [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+        <record id="survey_certifications_action" model="ir.actions.act_window">
+            <field name="name">Certifications Succeeded</field>
+            <field name="res_model">survey.user_input</field>
+            <field name="view_mode">tree</field>
+            <field name="context">{'search_default_quizz_passed': 1}</field>
+        </record>
+
+        <record id="res_partner_view_form" model="ir.ui.view">
+            <field name="name">res.partner.certification.stat.button</field>
```
Comment (tde-banana-odoo) [inline addons/website_slides_survey/views/res_partner_views.xml]: Guidelines: res.partner.view.form.inherit.slides.survey

File: addons/website_slides/views/slide_channel_partner_views.xml (hunk 13 above)
Comment (pko-odoo) [reply addons/website_slides/views/slide_channel_partner_views.xml]: Indeed, I implemented another view for slide.channel.partner and forgot to remove that line after changing it :)

Comment (pko-odoo) [reply addons/website_slides/views/slide_channel_partner_views.xml]: done !

**Comment by tde-banana-odoo** [comment]
@robodoo r+
